File: ripperdoc/wetware/ghost/scripts/pulse.py
#!/usr/bin/env python3

import base64
import json
import os
import sys
from pathlib import Path

PLUGIN_ROOT = Path(
    os.environ.get("CLAUDE_PLUGIN_ROOT") or os.environ.get("PLUGIN_ROOT") or str(Path(__file__).resolve().parent.parent)
)
GHOST_SENTINEL = Path.home() / ".claude" / ".cache" / ".ghost"
DEFAULT_EVENT = "UserPromptSubmit"


def read_source(dat_name: str, md_name: str) -> str | None:
    dat = PLUGIN_ROOT / dat_name
    if dat.exists():
        return base64.b64decode(dat.read_bytes()).decode()
    md = PLUGIN_ROOT / md_name
    if md.exists():
        return md.read_text()
    return None


def main() -> int:
    if not GHOST_SENTINEL.exists():
        GHOST_SENTINEL.parent.mkdir(parents=True, exist_ok=True)
        GHOST_SENTINEL.touch()

    content = read_source("pulse.dat", "PULSE.md")
    if not content:
        print("{}")
        return 0

    event_name = DEFAULT_EVENT
    try:
        payload = json.loads(sys.stdin.read() or "{}")
        if isinstance(payload, dict):
            name = payload.get("hook_event_name")
            if isinstance(name, str) and name:
                event_name = name
    except (json.JSONDecodeError, OSError):
        pass

    print(
        json.dumps(
            {
                "hookSpecificOutput": {
                    "hookEventName": event_name,
                    "additionalContext": content.strip(),
                }
            }
        )
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: ripperdoc/wetware/ghost/scripts/pulse.ts
#!/usr/bin/env -S preemdeck-bun
/**
 * pulse.ts — UserPromptSubmit persona-pulse injector.
 *
 * Reads the pulse source (base64 `pulse.dat` preferred, else `PULSE.md`) from the
 * plugin root and injects its stripped body via lib/hook.ts. Missing/empty is a
 * silent `{}` no-op. Default event UserPromptSubmit; stdin wins.
 */

import { readFile } from "node:fs/promises"
import { dirname, join } from "node:path"
import { exists } from "../../../../lib/fs.ts"
import { runInjectionHook } from "../../../../lib/inject.ts"

const DEFAULT_EVENT = "UserPromptSubmit"

/** The plugin root: CLAUDE_PLUGIN_ROOT || PLUGIN_ROOT || the script dir's parent. */
export const pluginRoot = (): string => {
  return process.env.CLAUDE_PLUGIN_ROOT || process.env.PLUGIN_ROOT || dirname(import.meta.dir)
}

/**
 * Read the pulse source: base64 `.dat` if present (decoded), else the plain
 * `.md`, else null.
 */
export const readSource = async (root: string, datName: string, mdName: string): Promise<string | null> => {
  const dat = join(root, datName)
  if (await exists(dat)) {
    return Buffer.from((await readFile(dat)).toString("utf8"), "base64").toString("utf8")
  }
  const md = join(root, mdName)
  if (await exists(md)) {
    return await readFile(md, "utf8")
  }
  return null
}

if (import.meta.main) {
  const root = pluginRoot()
  const content = await readSource(root, "pulse.dat", "PULSE.md")
  await runInjectionHook({
    event: DEFAULT_EVENT,
    render: () => {
      if (!content) return null
      return content.trim()
    },
  })
  process.exit(0)
}
